Consume newer Roslyn APIs for language services

drewnoakes opened the pull request:
Fixes https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1881089

Roslyn recently introduced async API for obtaining and releasing "batch" objects for applying updates in https://github.com/dotnet/roslyn/pull/72424.

This change increases the package versions, uses the newer API, fixes some obsolete usages, and gets things building by adding a few package references in order to break the tie on some assembly version conflicts during build.

The package `Microsoft.VisualStudio.Internal.MicroBuild.Swix` was removed, as it conflicted with behaviour inherited via `microsoft.visualstudioeng.microbuild.plugins.swixbuild`, which now comes in transitively, resulting in errors running `swc`. Having both loaded was doubling up the items used to populate arguments, and it was throwing _An item with the same key has already been added._
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/project-system/pull/9455)

ToddGrun commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/Workspace.cs:
```diff
@@ -525,16 +526,16 @@ async Task ApplyInBatchAsync()
-
                 try
                 {
+                    IAsyncDisposable disposableBatchScope = await Context.CreateBatchScopeAsync(CancellationToken.None);
```
>CancellationToken.None [](http://example.com/codeflow?start=96&length=22)

Do you remember why we didn't use the cancellationToken available to us here?

drewnoakes replied:
No, I believe we should pass it. Pushing an update.

tmeschter commented on src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/LanguageServices/Workspace.cs:
```diff
@@ -525,16 +526,16 @@ async Task ApplyInBatchAsync()
                 {
+                    IAsyncDisposable disposableBatchScope = await Context.CreateBatchScopeAsync(cancellationToken);
+
+                    await using ConfiguredAsyncDisposable _ = disposableBatchScope.ConfigureAwait(false);
```
Do we need the `ConfigureAwait(false)`? We should already be on a background thread.

drewnoakes replied:
Yeah I think it can go.

tmeschter commented on tests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests.csproj:
```diff
@@ -9,6 +9,8 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <PackageReference Include="StreamJsonRpc" />
```
Why do we need `StreamJsonRpc` everywhere?

drewnoakes replied:
This is what I hinted at with:

> gets things building by adding a few package references in order to break the tie on some assembly version conflicts during build.

Without this, restore completes successfully but the build fails with:

```
Found conflicts between different versions of "StreamJsonRpc" that could not be resolved.
There was a conflict between "StreamJsonRpc, Version=2.16.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" and "StreamJsonRpc, Version=2.18.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
    "StreamJsonRpc, Version=2.16.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was chosen because it was primary and "StreamJsonRpc, Version=2.18.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" was not.
    References which depend on "StreamJsonRpc, Version=2.16.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" [D:\NuGetPackages\streamjsonrpc\2.16.41\lib\netstandard2.0\StreamJsonRpc.dll].
        D:\NuGetPackages\streamjsonrpc\2.16.41\lib\netstandard2.0\StreamJsonRpc.dll
          Project file item includes which caused reference "D:\NuGetPackages\streamjsonrpc\2.16.41\lib\netstandard2.0\StreamJsonRpc.dll".
            D:\NuGetPackages\streamjsonrpc\2.16.41\lib\netstandard2.0\StreamJsonRpc.dll
    References which depend on or have been unified to "StreamJsonRpc, Version=2.18.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a" [].
        D:\NuGetPackages\microsoft.servicehub.framework\4.4.23\lib\net472\Microsoft.ServiceHub.Framework.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.servicehub.framework\4.4.23\lib\net472\Microsoft.ServiceHub.Framework.dll".
            D:\NuGetPackages\microsoft.servicehub.framework\4.4.23\lib\net472\Microsoft.ServiceHub.Framework.dll
            D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.query\17.8.53-pre\lib\netstandard2.0\Microsoft.VisualStudio.ProjectSystem.Query.dll
            D:\NuGetPackages\microsoft.visualstudio.rpccontracts\17.10.3-preview\lib\netstandard2.0\Microsoft.VisualStudio.RpcContracts.dll
            D:\NuGetPackages\microsoft.visualstudio.editor\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Editor.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.15.0\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.15.0.dll
            D:\NuGetPackages\microsoft.visualstudio.hotreload.components\6.0.0-preview.6.21380.4\lib\net472\Microsoft.VisualStudio.HotReload.Components.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.vs\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.VS.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.design\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Design.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.framework\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Framework.dll
            D:\NuGetPackages\microsoft.visualstudio.utilities\17.7.36307-preview.2\lib\net472\Microsoft.VisualStudio.Utilities.dll
            D:\NuGetPackages\microsoft.visualstudio.imaging\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Imaging.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.dll
            d:\repos\project-system\artifacts\Debug\bin\Dlls\net472\Microsoft.VisualStudio.ProjectSystem.Managed.dll
        D:\NuGetPackages\microsoft.visualstudio.language\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Language.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.visualstudio.language\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Language.dll".
            D:\NuGetPackages\microsoft.visualstudio.language\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Language.dll
            D:\NuGetPackages\microsoft.visualstudio.editor\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Editor.dll
            D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
        D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll".
            D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
        D:\NuGetPackages\microsoft.visualstudio.projectsystem.query\17.8.53-pre\lib\netstandard2.0\Microsoft.VisualStudio.ProjectSystem.Query.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.visualstudio.projectsystem.query\17.8.53-pre\lib\netstandard2.0\Microsoft.VisualStudio.ProjectSystem.Query.dll".
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.query\17.8.53-pre\lib\netstandard2.0\Microsoft.VisualStudio.ProjectSystem.Query.dll
        D:\NuGetPackages\microsoft.visualstudio.rpccontracts\17.10.3-preview\lib\netstandard2.0\Microsoft.VisualStudio.RpcContracts.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.visualstudio.rpccontracts\17.10.3-preview\lib\netstandard2.0\Microsoft.VisualStudio.RpcContracts.dll".
            D:\NuGetPackages\microsoft.visualstudio.rpccontracts\17.10.3-preview\lib\netstandard2.0\Microsoft.VisualStudio.RpcContracts.dll
            D:\NuGetPackages\microsoft.visualstudio.editor\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Editor.dll
            D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.query\17.8.53-pre\lib\netstandard2.0\Microsoft.VisualStudio.ProjectSystem.Query.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.15.0\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.15.0.dll
            D:\NuGetPackages\microsoft.visualstudio.hotreload.components\6.0.0-preview.6.21380.4\lib\net472\Microsoft.VisualStudio.HotReload.Components.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.vs\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.VS.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.design\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Design.dll
            D:\NuGetPackages\microsoft.visualstudio.utilities\17.7.36307-preview.2\lib\net472\Microsoft.VisualStudio.Utilities.dll
            D:\NuGetPackages\microsoft.visualstudio.imaging\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Imaging.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.dll
            d:\repos\project-system\artifacts\Debug\bin\Dlls\net472\Microsoft.VisualStudio.ProjectSystem.Managed.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.framework\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Framework.dll
        D:\NuGetPackages\microsoft.visualstudio.utilities\17.7.36307-preview.2\lib\net472\Microsoft.VisualStudio.Utilities.dll
          Project file item includes which caused reference "D:\NuGetPackages\microsoft.visualstudio.utilities\17.7.36307-preview.2\lib\net472\Microsoft.VisualStudio.Utilities.dll".
            D:\NuGetPackages\microsoft.visualstudio.utilities\17.7.36307-preview.2\lib\net472\Microsoft.VisualStudio.Utilities.dll
            D:\NuGetPackages\microsoft.visualstudio.imaging\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Imaging.dll
            D:\NuGetPackages\microsoft.visualstudio.languageservices\4.11.0-1.24225.10\lib\net472\Microsoft.VisualStudio.LanguageServices.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.dll
            d:\repos\project-system\artifacts\Debug\bin\Dlls\net472\Microsoft.VisualStudio.ProjectSystem.Managed.dll
            D:\NuGetPackages\microsoft.visualstudio.projectsystem.vs\17.8.53-pre\lib\net472\Microsoft.VisualStudio.ProjectSystem.VS.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.15.0\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.15.0.dll
            D:\NuGetPackages\microsoft.visualstudio.hotreload.components\6.0.0-preview.6.21380.4\lib\net472\Microsoft.VisualStudio.HotReload.Components.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.design\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Design.dll
            D:\NuGetPackages\microsoft.visualstudio.shell.framework\17.3.32804.24\lib\net472\Microsoft.VisualStudio.Shell.Framework.dll
            D:\NuGetPackages\microsoft.visualstudio.editor\17.0.32-g9d6b2c6f26\lib\net472\Microsoft.VisualStudio.Editor.dll
```

That error occurred for every project I added an explicit package reference to `StreamJsonRpc`.

drewnoakes commented:
Adding notes here as I investigate the SWIX conflicts that fall out of these package upgrades:

The error I see is:

```
  Unhandled Exception: System.ArgumentException: An item with the same key has already been added.
     at System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
     at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
     at System.Collections.Generic.Dictionary`2.Add(TKey key, TValue value)
     at WixToolset.Simplified.SimplifiedWixCompiler.Run(Messaging messaging)
     at WixToolset.Simplified.SimplifiedWixCompiler.Main(String[] args)
D:\NuGetPackages\microsoft.devdiv.tasks2\17.4.0-preview-2-32803-137\build\swix.targets(154,5): error MSB6006: "swc.exe" exited with code -532462766.
```

I believe that exception is coming from [here](https://github.com/wixtoolset/wix4-archive/blob/ec7c93b2c8c278c3e4e456ed65d624e40101c2e5/src/tools/swix/swc/SimplifiedWixCompiler.cs#L176) where "preprocessor defines" are being added.

Looking at the binlog shows the task does indeed have duplicates:

![image](https://github.com/dotnet/project-system/assets/350947/aca57066-436a-4e7e-824e-62f37e1e3eff)

Task logs its `CommandLineArguments` as:

```
CompileVsix:
  D:\NuGetPackages\microsoft.visualstudio.internal.microbuild.setuptooling\2.0.115\build\..\tools\bin\swc.exe
   -arch neutral
   -ext D:\NuGetPackages\microsoft.visualstudio.internal.microbuild.setuptooling\2.0.115\build\..\ext\DevDivSwixExtension.dll
   -ext D:\NuGetPackages\microsoft.visualstudio.internal.microbuild.setuptooling\2.0.115\build\..\tools\VsixSwixExtension.dll
   -o d:\repos\project-system\artifacts\Debug\VSSetup\Insertion\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles.vsix
   -d VisualStudioXamlRulesDir=d:\repos\project-system\artifacts\Debug\VSSetup\Rules\
   -d VisualStudioExtensionSetupDir=d:\repos\project-system\setup\
   -d Chip=AnyCPU
   -d MicrosoftPackageNamePrefix=Microsoft.
   -d "MicrosoftPublisher=CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US"
   -d "VSDefaultProgramPath=Software\Microsoft\VisualStudio_[InstanceId]\Capabilities"
   -d VsDdeApplication=VisualStudio.15.0
   -d Lang=enu
   -d Codepage=1252
   -d Culture=en
   -d LangCountry=en-US
   -d LCID=1033
   -d Chip=AnyCPU
   -d MicrosoftPackageNamePrefix=Microsoft.
   -d "MicrosoftPublisher=CN=Microsoft Corporation, O=Microsoft Corporation, L=Redmond, S=Washington, C=US"
   -d "VSDefaultProgramPath=Software\Microsoft\VisualStudio_[InstanceId]\Capabilities"
   -d VsDdeApplication=VisualStudio.15.0
   -d Lang=enu
   -d Codepage=1252
   -d Culture=en
   -d LangCountry=en-US
   -d LCID=1033
   d:\repos\project-system\artifacts\Debug\VSSetup.obj\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles\swr\core.neutral.swr d:\repos\project-system\setup\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles\CommonFiles.swr d:\repos\project-system\setup\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles\ext.xproj.swr D:\NuGetPackages\microsoft.visualstudio.internal.microbuild.setuptooling\2.0.115\build\..\tools\VsixExternalFolders.swr
   -t vsix
```

Where `-d` is a define, and we see the duplicates.

These defines are coming from `PackagePreprocessorDefinitions` items in the build. Digging through the binlogs shows that two build files are reassigning the property:

1. D:\NuGetPackages\microsoft.visualstudio.internal.microbuild.swix\2.0.115\build\Microsoft.Wix4.Swix.Tools.cleaned.targets
2. D:\NuGetPackages\microsoft.visualstudioeng.microbuild.plugins.swixbuild\1.1.286\build\Microsoft.Wix4.Swix.Tools.targets

They have near-identical code. Indeed one appears to have been copied from the other. My assumption is that they're not intended to be used together.

I had previously tried removing `microsoft.visualstudio.internal.microbuild.swix`, which worked locally but did not work in CI.

So now my plan is to find out why we're bringing in the other (`microsoft.visualstudioeng.microbuild.plugins.swixbuild`) and see if we can prevent that.

drewnoakes commented:
`microsoft.visualstudioeng.microbuild.plugins.swixbuild` is imported from a NuGet package:

![image](https://github.com/dotnet/project-system/assets/350947/721b111d-3127-4315-b04a-769d27a8e6f7)

There's no mention of that package in the dependencies tree though. Looking in the assets file (`D:\repos\project-system\artifacts\Debug\obj\Microsoft.VisualStudio.ProjectSystem.Managed.CommonFiles\project.assets.json`) also shows no direct mention of `microsoft.visualstudioeng.microbuild.plugins.swixbuild`, so I'm unclear on how this is being added and still ended up in the `.g.props` file.

Pushed another experimental commit.

drewnoakes commented:
That's making progress.

Now we have test failures. I assume this is due to updates to `Microsoft.VisualStudio.Threading`.

> Microsoft.VisualStudio.ProjectSystem.Properties.AssemblyInfoPropertiesProviderTests.SourceFileProperties_SetPropertyValueAsync line 184
>
>> Microsoft.VisualStudio.Threading.JoinableTaskContextException : An attempt to switch to the main thread failed to reach the expected thread. Was the JoinableTaskContext initialized on the wrong thread or with a SynchronizationContext whose Post method does not execute its delegate on the main thread?

Stack Trace: 

```
    MainThreadAwaiter.GetResult()
    SourceAssemblyAttributePropertyValueProvider.SetPropertyValueAsync(String value) line 102
    AssemblyInfoProperties.SetPropertyValueAsync(String propertyName, String unevaluatedPropertyValue, IReadOnlyDictionary`2 dimensionalConditions) line 102
    InterceptedProjectProperties.SetPropertyValueAsync(String propertyName, String unevaluatedPropertyValue, IReadOnlyDictionary`2 dimensionalConditions) line 80
    AssemblyInfoPropertiesProviderTests.SourceFileProperties_SetPropertyValueAsync(String code, String propertyName, String propertyValue, String expectedCode) line 191
```

There are test failures with `RenamerTests` too, along the same lines.

drewnoakes commented:
That exception is thrown here:

https://github.com/microsoft/vs-threading/blob/e4417b768162ac74b43e251320d11de771e7653b/src/Microsoft.VisualStudio.Threading/JoinableTaskFactory.cs#L854

I was able to fix these issues by suppressing the synchronization context for the tests in question.

drewnoakes commented:
@tmeschter @ToddGrun this is ready for review now.

ToddGrun commented:
The Roslyn interaction part LGTM

drewnoakes commented on tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Utilities/SynchronizationContextUtil.cs:
```diff
@@ -0,0 +1,28 @@
+
+        public void Dispose()
+        {
+            if (Interlocked.Exchange(ref old, null) is { } restoreMe)
```
Thus might not work correctly if the context was originally null. Simple fix.

tmeschter commented on Directory.Packages.props:
```diff
@@ -77,18 +82,18 @@
+    <PackageVersion Include="Microsoft.CodeAnalysis.Common"                                          Version="4.11.0-1.24225.10" />
     <PackageVersion Include="Microsoft.CSharp"                                                       Version="4.7.0" />
-    <PackageVersion Include="Microsoft.Net.Compilers.Toolset"                                        Version="4.10.0-3.24157.9" />
+    <PackageVersion Include="Microsoft.Net.Compilers.Toolset"                                        Version="4.11.0-1.24225.10" />
```
Note that this version doesn't need to match the versions of other Roslyn packages since it changes which compiler we use, rather than providing assemblies to reference. Not really relevant here but worth pointing out.

drewnoakes replied:
True. I made it the same so that we get into the habit of bumping them all together.

tmeschter commented on Directory.Packages.props:
```diff
@@ -26,18 +31,18 @@
     <PackageVersion Include="Microsoft.VisualStudio.Internal.MicroBuild.NpmPack"                     Version="2.0.117" />
     <PackageVersion Include="Microsoft.VisualStudio.Internal.MicroBuild.Swix"                        Version="2.0.115" />
     <PackageVersion Include="Microsoft.VisualStudio.Internal.MicroBuild.Vsman"                       Version="2.0.115" />
-    <PackageVersion Include="Microsoft.VisualStudioEng.MicroBuild.Core"                              Version="1.0.0" />
```
Do we no longer need this? Does it get brought in transitively?

drewnoakes replied:
Kind of. A different-but-similar package is brought in transitively. This was a head scratcher. More details in the PR description and comments below if you're interested.

tmeschter commented on tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/ProjectSystem/Properties/AssemblyInfoPropertiesProviderTests.cs:
```diff
@@ -183,6 +183,8 @@ public async Task ProjectFileProperties_GetEvaluatedPropertyAsync(string code, s
                     """[assembly: System.AssemblyDescriptionAttribute("MyDescription")]""")]
         public async Task SourceFileProperties_SetPropertyValueAsync(string code, string propertyName, string propertyValue, string expectedCode)
         {
+            using var _ = SynchronizationContextUtil.Suppress();
```
Comment about why we need to do this?

drewnoakes replied:
Adding comments to the `Suppress` method itself, as the reason is a bit long for a single-line comment here (and on other usages).

tmeschter commented on tests/Microsoft.VisualStudio.ProjectSystem.Managed.UnitTests/Utilities/SynchronizationContextUtil.cs:
```diff
@@ -0,0 +1,28 @@
+
+namespace Microsoft.VisualStudio;
+
+internal static class SynchronizationContextUtil
```
This could really use a doc comment explaining what it is, why we need it, and, if possible, guidance on when to use it.

drewnoakes replied:
I added a comment to the `Suppress` method that provides detail as to why I added this. Let me know if it's unclear or could be improved.

drewnoakes commented:
We are reverting this because it's broken build signing and is blocking other work. Will revisit.
